### ISSUE TITLE
Length encoding bug when combining dynamic and fixed-size arrays

### ISSUE THREAD

**Issue by christoph2806** [ethereumjs/ethereumjs-abi#22]
rawEncode doesn't handle mixed dynamic/fixed arguments correctly.
Example: 
    simpleEncode("someFunc(string,uint8[5])","xxxx",["1","1","1","1","1"]) 

should yield
"0x4a0fe00000000000000000000000000000000000000000000000000000000000000000**c0**0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000047878787800000000000000000000000000000000000000000000000000000000" 

but yields 
"0x4a0fe00000000000000000000000000000000000000000000000000000000000000000**40**0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000047878787800000000000000000000000000000000000000000000000000000000"

Proposed solution: 

replace rawEncode by 

```
    ABI.rawEncode = function (types, values) {
      var output = []
      var data = []
      var dynLength = []

      var headLength = 0

      for (var i in types) {
        var type = elementaryName(types[i])
        var value = values[i]
        var cur = encodeSingle(type, value)

        // Use the head/tail method for storing dynamic data
        if (isDynamic(type)) {
          output.push(0)
          data.push(cur)
          dynLength.push(cur.length)
          headLength += 32;
        } else {
          output.push(cur)
          dynLength.push(0)
          headLength += cur.length;
        }
      }

      for (var j=0; j< types.length; j++) {
          if (dynLength[j] > 0) {
              output[j] = encodeSingle('uint256', headLength);
              headLength += dynLength[j]
          }
      }

      return Buffer.concat(output.concat(data))
    }
```

**Comment by axic** [comment]
Thanks @christoph2806!

I think you're correct. The fixed length array shouldn't be counted as a dynamic length.

I'll look into this. Did you check this against other ABI implementations?

**Comment by Kuzirashi** [comment]
Hello it seems like a crucial thing for the library. Is there any status of this issue?

**Comment by holgerd77** [comment]
Hi @Kuzirashi, I'll try to have a look at this next week (sorry, we are still incredibly behind on working down bug reports in some parts of the EthereumJS ecosystem, we are getting better though).

**Comment by Kuzirashi** [comment]
This seems to be fixed by #60.

cc @holgerd77

**Comment by holgerd77** [comment]
@Kuzirashi Yes, thanks for reporting, will close this.

@christoph2806 Sorry for not directly considering the solution proposed here. This library was a bit in a stalled state for some time.
